### ISSUE TITLE
NoMethodError: undefined method `split' for nil:NilClass in extensiontask.rb:515:in `ruby_api_version'

### ISSUE THREAD

**Issue by kputnam** [rake-compiler/rake-compiler#156]
I'm not sure if I'm missing something obvious, but when running `rake native gem`... or even just `rake native`, it fails with

```
rake aborted!
NoMethodError: undefined method `split' for nil:NilClass
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-compiler-1.0.7/lib/rake/extensiontask.rb:515:in `ruby_api_version'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-compiler-1.0.7/lib/rake/extensiontask.rb:262:in `block in define_native_tasks'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:273:in `block in execute'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:273:in `each'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:273:in `execute'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
/Users/kputnam/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/monitor.rb:230:in `mon_synchronize'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:238:in `block in invoke_prerequisites'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:236:in `each'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:236:in `invoke_prerequisites'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:213:in `block in invoke_with_call_chain'
/Users/kputnam/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/monitor.rb:230:in `mon_synchronize'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:238:in `block in invoke_prerequisites'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:236:in `each'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:236:in `invoke_prerequisites'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:213:in `block in invoke_with_call_chain'
/Users/kputnam/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/monitor.rb:230:in `mon_synchronize'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/task.rb:183:in `invoke'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:160:in `invoke_task'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:116:in `each'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:116:in `block in top_level'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:125:in `run_with_threads'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:110:in `top_level'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:83:in `block in run'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:186:in `standard_exception_handling'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/lib/rake/application.rb:80:in `run'
/Users/kputnam/.rvm/gems/ruby-2.6.3/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/Users/kputnam/.rvm/gems/ruby-2.6.3/bin/rake:23:in `load'
/Users/kputnam/.rvm/gems/ruby-2.6.3/bin/rake:23:in `<main>'
/Users/kputnam/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'
/Users/kputnam/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'
Tasks: TOP => native => native:x86_64-darwin18 => native:stupidedi:x86_64-darwin18
```

I found another person reporting the same issue on [StackOverflow](https://stackoverflow.com/questions/54897269/rake-native-fails-with-undefined-method-split-for-nilnilclass). Like the person who wrote that, I'm confused how this is supposed to work. Where is `@ruby_versions_per_platform[platform]` supposed to be set? Should `ruby_api_version` expect `nil` as an argument? Am I doing something wrong when invoking `rake native gem`?

The StackOverflow post contains the info needed to reproduce the problem.

**Comment by kou** [comment]
Could you show how to reproduce this case?

**Comment by kou** [comment]
> The StackOverflow post contains the info needed to reproduce the problem.

Ah, OK.

**Comment by kou** [comment]
`rake native ...` isn't supported.
`rake native` is for `rake cross native ...`.

What do you want to do with `rake native`?

**Comment by kputnam** [comment]
I was following the example in https://github.com/rake-compiler/rake-compiler#generating-native-rubygems:

> Plus, rake-compiler tasks give you the extra functionality needed to build native gems by running:
>
> ```
> # rake native gem
> (... compilation output ...)
> ```

It looks like in that example, `rake native gem` would be the command to compile a binary gem for the current platform, while `rake cross native gem` would compile the binary gem for another platform.

I'm not sure what my precise goal is yet. I'm adding native code to my previously ruby-only library, and want to ensure the widest compatibility. It looks like pre-compiled binary gems are the way to go for Windows support (using rake-compiler's `rake cross native gem`?). But I'm not sure how to support UNIX-like hosts that don't have a compiler installed, apart from running `gem-compiler` on a similar host.

**Comment by kou** [comment]
Ah, it's outdated.

We need to build binaries for all available Ruby versions (2.4, 2.5 and 2.6 for now) to create a gem that works without C compiler. But the current `rake native gem` just builds a binary for the Ruby that we use now. If we use Ruby 2.6 to run `rake native gem`, the built gem only has a binary for Ruby 2.6.

And it's difficult to build "portable" binary for UNIX-like hosts. Because they may use different libc. If libc for build and libc for use is different, the binary will be crashed. We will use static linking to resolve it. But it has security issue. If libc has security problem, binary gem maintainers must update their gems ASAP. If we don't use static linking, user's system will update libc by its packaging system.

Will your gem have external dependency? For example, [sqlite3 gem](https://rubygems.org/gems/sqlite3) depends [SQLite3](https://www.sqlite.org/).

**Comment by bbuchalter** [comment]
I'm blocked by this error. Any advice on how we get past it?

**Comment by kputnam** [comment]
@kou Thanks for the response. Nope there aren't any external dependencies. I was a little ahead of myself because I haven't tried running my code on another platform, but at least now I've got some things compiling and running.

I want to support Windows, macOS, Linux all on 64-bit Intel and ARM. I guess there are a couple ways to go about that. The options I'm aware of (in order preference) are cross-compiling on a single build machine using rake-compiler, or using gem-compiler on a machine with the target OS/arch, or shipping source code and have users compile it locally.

**Comment by kou** [comment]
@bbuchalte could you show your use case?

**Comment by kou** [comment]
> I want to support Windows, macOS, Linux all on 64-bit Intel and ARM. I guess there are a couple ways to go about that. The options I'm aware of (in order preference) are cross-compiling on a single build machine using rake-compiler, or using gem-compiler on a machine with the target OS/arch, or shipping source code and have users compile it locally.

You need to use `rake cross native ...` instead of `rake native ...` for cross-compiling.

**Comment by pwoolcoc** [comment]
I'm running into this as well. My use-case is this: we have an internal gem that we want to include the compiled artifacts with when the gem is published. We'd like to do this because it means the gem won't have to re-build at `bundle install` time, we'd like to compile it on our CI, include it with the gem files, then be able to just pull it down and install it at install time. So we don't need cross-compilation or anything, we only need to support a single platform, and our runtime platform will be the same as our build platform

**Comment by kou** [comment]
You want to avoid rebuilding the internal gem for CI, right?

In the case, it's better that you use cache feature provided by CI. You can cache `~/.gem` or `vendor/bundle` then all gems including the internal gem are cached without any additional configuration in the internal gem and your CI setting.

**Comment by larskanis** [comment]
This is a regression caused by https://github.com/rake-compiler/rake-compiler/commit/0dc23504cb03ed2fb3c506e1bb58af48d3851d1e . #166 fixes this bug.

**Comment by orgads** [comment]
This is solved by #166 indeed.

When is the next release planned?

**Comment by kou** [comment]
I've released.
